internal/events/flagged_driver.go

```diff
@@ -0,0 +1,162 @@
+	experimentPub message.Publisher
+	baseSub       message.Subscriber
+	experimentSub message.Subscriber
+
+	flags             openfeature.IClient
+	publishedMessages metric.Int64Counter
+	readMessages      metric.Int64Counter
+}
+
+// Publish implements message.Publisher.  If the message is flagged by the
+// alternate_message_driver experiment, it is published to the experiment
+// driver, otherwise it is published to the base driver.
+func (f *flaggedDriver) Publish(topic string, messages ...*message.Message) error {
+	// Each message has its own context, so they _could_ be in different flag treatments
+	for _, m := range messages {
+		emptyContext := engcontext.EntityContext{}
+		if engcontext.EntityFromContext(m.Context()) == emptyContext {
+			zerolog.Ctx(m.Context()).Warn().Str("stack", string(debug.Stack())).Msg("No entity in context")
+		}
+		if flags.Bool(m.Context(), f.flags, flags.AlternateMessageDriver) {
+			if err := f.experimentPub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", true)))
+		} else {
+			if err := f.basePub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", false)))
+		}
+	}
+	return nil
+}
+
+// Subscribe implements message.Subscriber.  In this case, it should subscribe
+// to both the base and experiment drivers, because we might get messages for
+// either.
+func (f *flaggedDriver) Subscribe(ctx context.Context, topic string) (<-chan *message.Message, error) {
+	out := make(chan *message.Message)
+	base, err := f.baseSub.Subscribe(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to subscribe to base: %w", err)
+	}
+	experiment, err := f.experimentSub.Subscribe(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to subscribe to experiment: %w", err)
+	}
+	go func() {
+		defer close(out)
+		// Cribbed from https://medium.com/justforfunc/why-are-there-nil-channels-in-go-9877cc0b2308
+		for base != nil || experiment != nil {
+			select {
+			case msg, ok := <-base:
+				if !ok {
+					base = nil
+					continue
+				}
+				out <- msg
+				f.readMessages.Add(ctx, 1, metric.WithAttributes(attribute.Bool("experiment", false)))
+			case msg, ok := <-experiment:
+				if !ok {
+					experiment = nil
+					continue
+				}
+				out <- msg
+				f.readMessages.Add(ctx, 1, metric.WithAttributes(attribute.Bool("experiment", true)))
+			}
+		}
+	}()
+	return out, nil
+}
+
+// Close implements message.Publisher and message.Subscriber.  It closes all
+// the drivers managed by the flagged publisher.
+func (f *flaggedDriver) Close() error {
+	if err := f.basePub.Close(); err != nil {
+		return fmt.Errorf("Failed to close base publisher: %w", err)
+	}
+	if err := f.experimentPub.Close(); err != nil {
+		return fmt.Errorf("Failed to close experiment publisher: %w", err)
+	}
+	return nil
+}
```

Comment:
Do we also want to invoke the `baseCloser()` and `experimentCloser()`? (By storing them?)
I think we close everything in eventer, just wanted to highlight this:
```go
	return &eventer{
		router:            router,
		webhookPublisher:  pubWithMetrics,
		webhookSubscriber: subWithMetrics,
		closer: func() {
			//nolint:gosec // It's fine if there's an error as long as we close the router
			pubWithMetrics.Close()
			//nolint:gosec // It's fine if there's an error as long as we close the router
			subWithMetrics.Close()
			// driver close
			cl()
		},
		msgInstruments: metricInstruments,
	}, nil
```

Reply:
`makeFlaggedDriver` returns a `closer` function that calls the returned `baseCloser` and `experimentCloser`.

I'm not entirely sure why we have both `closer` functions and a `Close` method on publisher and subscriber, but if there's a difference, it feels like we shouldn't cross the streams.  (And if there's no difference, we should clean that up in a _separate_ refactor...)

Reply:
Yeah, valid point.